File: consumatore.c
#include <stdio.h>
#include <fcntl.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#define MAX_STRING_LENGTH 256

// consumatore.c e' un filtro
int main(int argc, char* argv[]){

	char *file_in, read_char, buf[MAX_STRING_LENGTH];
	int nread, fd,i ,j=0,k=0;
    char prefix[MAX_STRING_LENGTH];
	//controllo numero argomenti
	if (argc != 3 && argc!=2){ 
		perror(" numero di argomenti sbagliato"); exit(1);
	} 
	strcpy(prefix,argv[1]);
    //strcpy(prefix,argv[1]);
    if(argc == 3){
        
        file_in = argv[2];
        
        fd = open(file_in, O_RDONLY);
    
        if (fd<0){
            perror("P0: Impossibile aprire il file.");
            exit(2);
        }
    }
    else fd=0;
    printf("prefix = %s\n",prefix);
    
            int len_prefix=strlen(prefix);

	while(nread = read(fd, &read_char, sizeof(char))) /* Fino ad EOF*/{
		if(nread>=0){
                if(read_char==prefix[k] && k<len_prefix){
                    
                   /* if(read_char =='\n' || read_char==' ') {
                        memset(buf,0,strlen(buf));

                        k=0;
                    }*/
                    if(k+1>=len_prefix || read_char=='\n') k=0;
                    else {
                        buf[k]=read_char;
                        k++;
                    }
                }
                else if(read_char!=prefix[k] && k!=0){
                    buf[k]='\0';
                    k=0;
                    write(1, buf, sizeof(char)*strlen(buf));
                    
                    write(1, &read_char, sizeof(char));
                }
                else if(read_char!=prefix[k] && k==0){
                    k=0;
                    write(1, &read_char, sizeof(char));
                }
        }
		else{
			printf("(PID %d) impossibile leggere dal file %s", getpid(), file_in);
			perror("Errore!");
			close(fd);
			exit(3);
		}
	}
	
	close(fd);
}

File: Consumatore.java
package reties0;


	
	import java.io.BufferedReader;
import java.io.FileNotFoundException;
	import java.io.FileReader;
	import java.io.IOException;

	// Consumatore e' un filtro
	public class Consumatore {
		public static void main(String[] args) throws FileNotFoundException {
			FileReader r = null;
			char carattere;
			int x;
			String str;
			BufferedReader in= new BufferedReader(new FileReader(args[1]));
			
		
			if (args.length != 1 && args.length != 2){
				System.out.println("Utilizzo: consumatore <inputFilename>");
				System.exit(0);
			}
		  str = args[0];
			
			try {
				//in= new BufferedReader(new FileReader(args[1]));
				r = new FileReader(args[1]);
				
			} catch(FileNotFoundException e){
				System.out.println("File non trovato");
				System.exit(1);
			}
			
			try {
				
				
				while ((x = r.read()) >= 0) {
					carattere = (char) x;
					if(!str.contains(String.format("%c",carattere)))
						System.out.print(carattere);
				}
				r.close();
			} catch(IOException ex){
				System.out.println("Errore di input");
				System.exit(2);
			}
	}
	}

